### DISCUSSION TITLE
Bringing TypeScript to Carbon

### DISCUSSION THREAD

**Discussion by joshblack** [carbon-design-system/carbon#10752] [Ideas]
This is an umbrella discussion around bringing TypeScript to Carbon, specifically to `@carbon/react` but could be included in all of our packages that ship JavaScript. This will be a living discussion that will be updated over time.

### August 2023   _🎶 We're halfway there 🎶_

Actually we're past halfway:

* 97 issues have been completed ✅ 
* 71 issues remain

Current status can be found in the issue:

* https://github.com/carbon-design-system/carbon/issues/12513

The same info can be found in the [TypeScript Adoption project](https://github.com/orgs/carbon-design-system/projects/53). It contains an [insights chart](https://github.com/orgs/carbon-design-system/projects/53/insights/2) that shows progress over time.

<details>
  <summary><h3>Click to expand previous updates</h3></summary>
  
**January 2023**

The effort to add types to `@carbon/react` is now underway. See the [announcement below](https://github.com/carbon-design-system/carbon/discussions/10752#discussioncomment-4710978) for details

**May 20th, 2022**

There is a kick-off meeting for the working group on **May 23rd, 2022**. I will be capturing notes and sharing them below. The intent of this meeting is to:

- Establish a direction for integrating TypeScript and Carbon based on the options presented and availability of individuals
- Nominate individuals to help lead or drive this effort, ideally from both Carbon and the community

**April 18th, 2022**

A Slack channel was created for a TypeScript working group for IBMers, if you're interested in helping out with this workstream please join: `#carbon-wg-typescript`.

We are currently investigating the as-is, overall design, and discrete, incremental deliverables that we can accomplish as a group.
</details> 

## Overview

Currently, teams who use TypeScript and Carbon will have types defined either through the community-owned `@types/carbon-components-react` or define them manually. This experience is not currently ideal because there is a chance that these types can get out of sync with the code shipped in the latest release on `npm`.

In order to improve this experience, the team has several ways that we can address this issue:

- Commit to maintaining types on DefinitelyTyped and release them in lockstep with the release
- Commit to delivering types through the individual `npm` packages, either through:
  - Manually authoring types through a `*.d.ts` file
  - Authoring source code in TypeScript and publishing types from the compilation process

## Questions

- Which type strategy do we want to follow?
  - Author source in TypeScript
    - How long would this process take?
    - Would authoring our source in TypeScript prevent bugs that we typically encounter?
  - Manually author types and ship with package
    - One option could be to follow MUI's approach which generates prop types from a `*.d.ts` file
  - Manually author types and ship with DefinitelyTyped

## Next steps

Our team is currently underway shipping our v11 release in March. Starting in Q2, we'll be working towards migration but will begin taking on new work. In particular, our focus will likely be on new components, testing (specifically visual regression testing), and documentation through Storybook and on our website.

However, if the consensus is that we need to ship TypeScript bindings then we really would like to deliver on that. In the interim, here are some of the open questions that we would like to address:

- How important is TypeScript integration relative to other efforts?
- What is the scope of the integration between Carbon and TypeScript
  - What kind of semver expectations exist around types shipped as part of a package?
  - How long would it take to:
    - Author source files in TypeScript
    - Manually author definition files and have them generate `propTypes`
- How can we provide an experience that works for both TypeScript and non-TypeScript users, in particular for prop types?

## Links & Resources

- Previous discussions
  - https://github.com/carbon-design-system/carbon/issues/5873

**Comment by seanlaff** [comment]
Thanks for starting the discussion @joshblack. I can give some perspective from our journey on converting a codebase to typescript.

If typescript types live in this repo, I think it makes the most sense to allow the components to benefit form the robustness they provide, by converting underlying components to typescript.

Typescript can be adopted piecemeal
1. Typescript and non-typescript files can live side by side
2. The "specificity" of types can start weak and grow in their specificity over time, eg:

```ts
type Props = {
  direction: string,
  onSuccess: () => void,
  icon: any
}
```
can evolve to 
```ts
type Props = {
  direction: "up" | "down" | "left" | "right",
  onSuccess: () => void,
  icon: CarbonIcon
}
```

For these reasons I think typescript adoption could be a collective effort- one that slowly and steadily improves over time as code is refactored to typescript. It doesn't have to a be done in a "big-bang" major carbon release.

Those of us in IBM that use typescript + carbon have considered contributing to DefinitelyTyped but it's a lost cause because it will always be out of sync with the underlying carbon code. If the carbon team laid the groundwork for typescript support in this repo, I (and I'm sure others) would be happy to help contribute our time.

**Comment by nertzy** [comment]
I am an open source (non-IBM) user of Carbon and I have contributed some issues and pull requests here and there. My team is transitioning from JavaScript to TypeScript and the types from DefinitelyTyped have mostly been a great help during our migration. We have needed to route around the occasional discrepancy, but it hasn't been a major impediment.

That said, we are also some open source components (such as Tearsheet) from [Carbon for IBM Products](https://github.com/carbon-design-system/ibm-cloud-cognitive/), which doesn't have any type information at all. It's been more challenging to figure out what to do to temporarily work around the lack of formal types there.

I'm particularly interested in seeing this library become even more of an accelerator for our work by having types. It's really great to get so much functionality and also to be able to contribute back when we discover improvements that can be made.

I know it's probably a challenge for y'all to coordinate, but I wanted to add a different kind of voice to the conversation. Thanks!

**Comment by SimeonC** [comment]
We just upgraded to the v11 version of `@carbon/icons-react` and discovered there are no types for it yet in definitely typed. I wrote a small rough script to get the types done for that and thought I'd share it in this conversation as a starter for generating the types (at least for the main export which is enough for us)

```js
import fs from 'fs';
import path from 'path';

import carbonIcons from '@carbon/icons-react';

const fileContent = `${Object.keys(carbonIcons).reduce(
  (result, iconName) =>
    `${result}  declare export const ${iconName}: ComponentType;\n`,
  `/* this file is generated, recommend adding to prettier and eslint ignore and committing to git. DO NOT EDIT */
  declare module '@carbon/icons-react' {
    type SizeType = 16 | 20 | 24 | 32;
    type ComponentType = React.ForwardRefExoticComponent<
      {
        size: SizeType | \`\${SizeType}\`;
      } & React.RefAttributes<SVGSVGElement>
    >;
`
)}\n}`;

fs.writeFileSync(
  path.resolve('./src/definitions/carbonIcons.d.ts'),
  fileContent
);
```

**Reply by metonym** [reply]
`@types/carbon__icons-react` has been updated to [v11.5.0](https://www.npmjs.com/package/@types/carbon__icons-react).

You should now see the correct types for v11 icons.

<img width="600" alt="@types:carbon__icons-react@11 5 0" src="https://user-images.githubusercontent.com/10718366/182986702-6fda0202-d640-49d7-8cb4-10906eb227ea.png">

**Reply by soukainakhalkhouli** [reply]
@SimeonC We just upgraded to the v11 version of @carbon/react  and discovered there are no types for it yet in definitely typed. Have you never create a script to get the types done for that and thought .Could you please share it with me ?

**Reply by SimeonC** [reply]
The script we're using at the moment is here https://github.com/tablecheck/tablecheck-react-system/blob/main/packages/scripts/scripts/utils/writeCarbonIconTypes.js

Auto generating of types means we don't have to wait for the types package to be updated.

**Reply by kosdziuba** [reply]
I recently went through [the guide for Angular](https://carbondesignsystem.com/developing/angular-tutorial/step-1) (Angular version 15.2.0) and had a problem with types for `@carbon/icons` (version 11.17.0).

Solved it by using an adapted version of the script that @SimeonC provided.

**Comment by erifsx** [comment]
As a long time user of carbon-components-angular, which natively supports Typescript out of necessity, who has just recently taken the plunge into carbon-components-react for a new project within IBM, I can't overstate how many times having a typed component system has saved me from bugs over the years, and I think would be an extremely valuable addition to the carbon-components-react library.  

While I understand, it's outside the purview of a component library, it opens the possibility for consumers to type the entire interaction chain from network call to rendering thereby catching, at compile time, possible typos that result from changes to a backend.  To a full stack developer, that kind of error catching during refactorings is invaluable.

I'd echo what @seanlaff had mentioned in that typing doesn't need to immediately be all encompassing.  The carbon-angular team would be a probably be a great resource for figuring out an LOE with regards to type maintenance.

Our larger concern at the moment is that our current react application is based on the definitely typed `@types/carbon-components-react` typings, but given this move to the new `@carbon/react` package, that new packaging scheme is completely inaccessible to us until those typings exist from some source.

**Reply by erifsx** [reply]
Just to add to that, adding type definitions to a package will also be a boon to developers who don't use or plan to use typescript at all.  

Most modern IDEs/Editors (VSCode/ Webstorm) will look at type definitions in a package to provide automatic code completion, even to a project that's entirely javascript.  Below is a minimal example in VSCode:
![image](https://user-images.githubusercontent.com/997572/163477089-4c580553-24d1-4acc-bb33-15a486dba498.png)
![image](https://user-images.githubusercontent.com/997572/163477142-3cd27c1d-c686-4938-ab95-b79f7a73c8cd.png)
Note that there's no tsc, webpack or tsconfig anywhere:
![image](https://user-images.githubusercontent.com/997572/163474067-1a2e4fdf-a7f1-49c8-86c8-6df2abffe958.png)

**Reply by jrplacer** [reply]
To echo what most have already said here, I think that supporting TypeScript would be mutually beneficial to both the carbon development community and to the carbon developers.  To be honest in my opinion this would greatly reduce development times, improve quality, and improve the overall development experience for both consumers of Carbon and for your developers developing Carbon.  Using libraries without TypeScript after using TypeScript for many years now is a very frustrating experience, it feels like going back 10 years.  @joshblack As an IBMer I offered to help the Carbon team late last year with supporting TypeScript as a few of my team members were struggling with out of date typings from definitely typed, but was essentially brushed off.  Find me on slack if you are serious about it.  We would love to help get involved.

**Reply by joshblack** [reply]
Hey there @jeffplacer!

I'm sorry that it felt like the team was brushing you off, especially when you were offering to help. 

I know personally I'm definitely interested in improving the support of TypeScript and will connect with you and others who are offering help 👍 I'll spin up a working group for this where we can chat about the logistics.

**Comment by joshblack** [comment]
Just wanted to post a quick update to this discussion 👀 

---

I just added an "Updates" section to the original discussion, this should include a rough timeline of where this discussion is at and what is currently happening.

The first update added to this section was a notice around us creating a working group over on Slack for IBMers to join. If you're interested in helping out at all, please join `#carbon-wg-typescript`. If you're external to IBM and would like to help out feel free to leave a comment here.

In this group, we'll be starting out with:

- Understanding the as-is experience for teams using TypeSCript
- Document known design goals for this effort
- Decide on discrete, incremental deliverables that we can accomplish relative to this effort

If you have any questions, feel free to reach out or leave a comment here!

**Reply by kubijo** [reply]
I was just about to post here that I would like to help as well when time and employment permit...

**Reply by jaclynRaine** [reply]
I'm interested in joining the slack and helping when I can if that's still an option!

**Comment by rampadc** [comment]
I've been waiting for Typescript to be added to carbon and had been told that it will never make it into the roadmap. So great thanks to @joshblack for starting the convo. Carbon is highly adopted. It's the design system IBM uses for its recent products. Typescript is self-documenting. It gives developers a well-defined contracts, an API, to what a component can do. Providing typed interfaces reduces the friction of having to write documentation by hand and greatly improves the developer experience. As others mentioned above, if I don't know how to use something, I'd hit Ctrl/Cmd+Space and get a list of options.

### Adopting Typescript

**How important is TypeScript integration relative to other efforts?**

> Carbon v11 focuses on quality of life updates for designers and developers

As a developer, I'd like to have some quality of life updates please. Seeing a sea of red squiggly lines for having to mix Typescript (web framework) and Javascript (Carbon React) together brings me great pain.

**Which type strategy do we want to follow?** 

I vote for authoring the source in Typescript. The developer who develops the components knows best what the component should offer. Any additional documentation can be added in JSDoc inline with the Typescript code. 

Adding types later in DefinitelyTyped usually means backtracking through design notes if the same developer is to write the types. If not, then the type writer would have to reach out to the original developer for the meaning behind a variable. You get the whole "who's going to maintain the types?" and "why are we doing duplicated work?" type questions.

**How long would this process take?** 

This is like asking how long is a piece of string. How long does it take you to decipher your old code and add in types? Have the developers ever worked with Typescript before?

 **Would authoring our source in TypeScript prevent bugs that we typically encounter?**

Same argument as above - the developer developing the component benefits from type checking.

Carbon is highly adopted. It's the design system IBM uses for its recent products. Hopefully, having the core Carbon team adopting Typescript would inspire/force downstream IBM teams to also adopt Typescript in their Carbon components. Most of IBM components I've used are lightly documented. I had to spend hours figuring out what settings are available and scour through examples (if there were any) to learn how to use a component. Offering Typescript improves the end developer experience of using Carbon in their projects.

**Comment by joshblack** [comment]
## Monday, May 23rd, 2022

Below is the agenda for the initial Working Group meeting. This post will also be updated by me to include notes from the session, including next steps or key decisions.

**Agenda**

- Introductions
- Understanding the as-is
  - How do users currently access TS types for Carbon libraries?
- Understanding the to-be
  - Where can we go from here to successfully deliver TS types to users?
  - Breaking down each option presented thus far
- Building consensus
  - Is there an option that resonates for people in the Working Group?
- Establishing a timeline and key contributors
  - How long might it take us to complete each option? 
  - Who from Carbon and from the community is available to help drive and organize the working group towards this effort?
  - When should this group meet next to talk about progress?
- Discussion, Q&A

## Understanding the as-is

- Roughly 25% of teams that use Carbon are also using TypeScript
  - As a result, roughly 75% of the existing user base is using propTypes
- TypesScript types currently are only available from DefinitelyTyped through `@types/carbon-components-react`
- TypeScript types are not available for `@carbon/react`
- TypeScript types can be out of sync with the latest Carbon release
- TypeScript types may be incorrect or may not completely cover the prop types a component may have
- TypeScript types from DefinitelyTyped may need to be overridden or changed locally for a project
- Contributing types to DefinitelyTyped does not feel worthwhile to product teams

## Understanding the to-be

At a high level, the decision for TypeScript types and Carbon boils down to either authoring the types alongside the source or authoring in DefinitelyTyped. This can be further broken down into:

- Authoring types alongside source
  - Authoring definition files
  - Authoring source code in TypeScript
- Authoring in DefinitelyTyped

| Strategy         | Duration | Maintainer        |
| :--------------- | :------- | :---------------- |
| Definition files | Medium   | Carbon, Community |
| Source code      | Long     | Carbon, Community |
| DefinitelyTyped  | Short    | Community, Carbon |

<details>
  <summary>Package ecosystem</summary>

- `@carbon/react` (aka `carbon-components-react`)
- `@carbon/icons-react`
- `@carbon/icons`
- `@carbon/icon-helpers`
- `@carbon/feature-flags`
- `@carbon/pictograms-react`
- `@carbon/pictograms`
- `@carbon/colors`
- `@carbon/layout`
- `@carbon/motion`
- `@carbon/themes`
- `@carbon/type`
</details>

**Challenges or unknowns**

> Authoring the source in TypeScript would require a significant time investment and would require teams to wait until that process is complete. Are there incremental approaches that we could follow?

> When authoring the source in TypeScript, how do we ensure parity between prop types and TypeScript types?

As maintainers, will you need to update the definition twice (once in TS, once in propTypes) or is there a way to automatically keep them in sync?

> What best practices should we be aware of for authoring types for a library? How do we make sure we avoid needing to make breaking changes in types?

> What best practices should we be aware of for authoring types that work against multiple versions of TypeScript?

**Incremental approach**

It may be possible for us to find ways to incrementally introduce TypeScript into the monorepo, starting with icons due to https://github.com/carbon-design-system/carbon/issues/11317. This would allow us to adopt it for a subset of packages (`@carbon/icons-react`, `@carbon/icon-helpers`, and maybe `@carbon/icons`) instead of having to take on the whole ecosystem at once.

Similarly, we could try and author the definition files in the source code without changing the underlying source code to TS (yet). Over time, the idea would be to shift types from being written manually to coming from the underlying source code. To the end-user, there should be no noticeable impact (hopefully). This would allow TS users to adopt sooner and would help with contributions to the typing files themselves as users would be contributing to the project directly.

## Discussion, Q&A

**Reply by joshblack** [reply]
**Next steps**

- [x] Playback meeting to the Carbon team, get feedback on direction (DefinitelyTyped or bring into monorepo) @joshblack 
- [x] Reach out in the #carbon-wg-typescript channel for contributors @joshblack 
- [x] Schedule next meeting in ~2 weeks @joshblack

**Comment by kubijo** [comment]
> When authoring the source in TypeScript, how do we ensure parity between prop types and TypeScript types?

I have used [`babel-plugin-typescript-to-proptypes`](https://github.com/milesj/babel-plugin-typescript-to-proptypes) for quite some time now in our project ... hence we are not authoring prop-types in TypeScript components at all

**Reply by joshblack** [reply]
Thanks @kubijo! Will check it out 👀

**Comment by joshblack** [comment]
## Monday, June 6th 2022

This is the second meeting of the TypeScript Working Group.

**Agenda**

- Discuss playback of TS WG findings to the Carbon team
- Go over the initial PR for adding types to `@carbon/icons-react`
- Talk about the contribution process for contributors on GitHub
- Discuss next steps

### Playback

Overall, I believe that the team is receptive to trying this out and seeing how it can improve the experience of authoring and using Carbon.

There were some concerns or questions raised, including:

- Would the move to TypeScript make it harder for an individual to contribute?
- Do packages that we depend on ship types (or has an `@types` package that is suitable)
- How does supporting multiple versions of TypeScript or a library work? For example:
  - Supporting React 17 and React 18 if the types differ
  - Supporting different TypeScript versions if a feature has been changed in some way
- What kind of guardrails could we put in place to make sure we don't ship breaking changes in our TS types in a minor release?

### Initial TypeScript Pull Request

Link: https://github.com/carbon-design-system/carbon/pull/11533

Feedback:

- What would be the ideal TypeScript config for the project?
- Is the proposed setup suitable for a monorepo?
- Are things like CI Checks appropriately set up?

### Contribution process

What would it look like for people wanting to get involved? How do we want to self-organize? How should we structure our work to make it easy to iterate on ideas?

### Next steps

What does a good first deliverable for this group look like?

**Questions**

- What happens if a package begins shipping with types but an `@types` package is currently installed on the project? Does one take precedence over the other?

### Q&A

**Reply by joshblack** [reply]
**Next steps**

- [x] Create a TypeScript Working Group team and add contributors to it (@joshblack)
- [x] Create an outline of the first deliverable (using the DefinitelyTyped types with Storybook) (@joshblack)
  - [x] Once this is created, post in the Working Group Slack channel for feedback and for more details from contributors
- [x] Add a section to our next Working Group meeting where we go over the monorepo and common processes/practices with it (@joshblack)
- [x] Share any resources in the meantime that go over the monorepo structure in the Working Group Slack channel (@joshblack)

**Reply by SimeonC** [reply]
> What happens if a package begins shipping with types but an @types package is currently installed on the project? Does one take precedence over the other?

The `@types` package takes precedence and needs to be uninstalled under normal/usual setups.

**Comment by joshblack** [comment]
## Monday, June 27th, 2022

**Agenda**

- Go over pending action items for members
- Monorepo overview
- Status of [initial Storybook migration](https://github.com/carbon-design-system/carbon/issues/11587)

### Pending action items

For members who have volunteered their time to help out, we have an official team on GitHub that youc an join: https://github.com/orgs/carbon-design-system/teams/wg-typescript

You should have received an invite for this team if you shared your GitHub username on Slack, if you have not yet received an invite or need an updated one let me know!

### Monorepo overview

This section will be a brief overview of the monorepo, its structure, and common tasks you may need to perform when working within it.

### Storybook migration

We currently have the following issue in scope for the working group: https://github.com/carbon-design-system/carbon/issues/11587

This time slot is to play back where that workstream is currently at and see if there is any additional participation from the WG that could help. move it forward.

**Reply by salvoravida** [reply]
@joshblack https://github.com/orgs/carbon-design-system/teams/wg-typescript
is it open to community contributors?

**Reply by joshblack** [reply]
Hey @salvoravida! 👋 Definitely, I think we'd love to have anyone interested participate.

**Reply by salvoravida** [reply]
Anyway, I get a 404

**Reply by lidongyang-Ethan** [reply]
yes I get a 404.is it open to community contributors? @joshblack

**Reply by tay1orjones** [reply]
@salvoravida @lidongyangLeo you should have received invites just now

**Comment by patelnav** [comment]
@joshblack any updates on this? Also can't access the team you mentioned above

**Reply by nickmagruder** [reply]
Also very interested in any updates! The link to the team is giving a 404: https://github.com/orgs/carbon-design-system/teams/wg-typescript

Also, would it be possible to get added to the Slack? Thanks!

**Reply by tay1orjones** [reply]
Hey there! There hasn't been much movement since the previous check in. Current status is the groundwork is mostly laid to get a PR up for https://github.com/carbon-design-system/carbon/issues/11587. We have a check in meeting scheduled for today I can add you to if you provide me your email.

The Slack is internal-only unfortunately, but I can spin up a channel on [our discord](https://discord.gg/KAECRaWdzE).

**Reply by tay1orjones** [reply]
@patelnav @nickmagruder you should have received invites just now

**Reply by jaclynRaine** [reply]
@tay1orjones can you add me as well, please?

**Comment by tay1orjones** [comment]
## Tuesday, August 23rd, 2022

**Agenda**

- New primary contact
- Go over pending action items for members
- External contributors
- Status of [initial Storybook migration](https://github.com/carbon-design-system/carbon/issues/11587)
- Monorepo pull request overview

### New primary contact

@tay1orjones 👋 😃 

If you're interested in helping to coordiate the working group, please let me know. This could include:

* Scheduling check ins
* Participating in slack/discord threads
* Review PRs
* Open/organize issues

### Pending action items

For members who have volunteered their time to help out, we have an official team on GitHub that you can join: https://github.com/orgs/carbon-design-system/teams/wg-typescript

You should have received an invite for this team if you shared your GitHub username on Slack, if you have not yet received an invite or need an updated one let me know!

### External contributors

We'll be setting up a channel on our [discord](https://discord.gg/KAECRaWdzE) for external contributors. We'll post updates there in addition to this discussion. In the future, our check ins can include those who are in the discord channel - please DM me an email to add you to the invite.

### Storybook migration

We currently have the following issue in scope for the working group: https://github.com/carbon-design-system/carbon/issues/11587

This time slot is to play back 
* Where the workstream is currently at
* Determine what's needed to get a PR up, identify any blockers
* See if there is any additional participation from the WG that could help move it forward. 
* Next steps once the PR is in, how to coordinate work

### Monorepo pull request overview

This section will be a brief overview of the pull request process for the monorepo.

**Reply by tay1orjones** [reply]
In the meeting today we discussed current status and some next steps: 

- [x] @mbarrer's going to put up a PR laying the groundwork for adding types to the repo
- [ ] Once that’s in, each component in the repo will need updated. I’ll make an issue with a checklist for each component. At this point we’d love to have anyone and everyone pitch in to make these updates!
- [x] I'll make a channel in [discord](https://discord.gg/C89hy2Ch) for the working group 

The overall focus moving forward is to get the repo to a place where DT types are ported over to be included as part of the package. Some DT types are outdated and will need updated. Some components do not have DT types and will need added. 

  If you’re interested in helping out and taking on a component, respond in this thread and I’ll include you on the checklist issue once it’s up.

**Comment by tay1orjones** [comment]
I've created the [`#wg-typescript` channel](https://discord.com/channels/689212587170201628/1011662356977287219) on discord.

Also the PR laying the groundwork for adding types is up! https://github.com/carbon-design-system/carbon/pull/12000

**Comment by br4e** [comment]
I am a heavy user of Typescript and Carbon. Unfortunately if Carbon 11 does not support typescript, we can not upgrade our repos to take advantage of the other quality of life updates in 11. Upgrading without typescript would be a significant quality of life decrease overall.

**Comment by br4e** [comment]
However, thank you for building carbon! It's a gamechanger.

**Comment by marekdano** [comment]
When do you think that TS can be implemented in carbon 11?

**Comment by tay1orjones** [comment]
## Status Update

The foundations of two different approaches to providing types are up in PRs right now for consideration:

1. https://github.com/carbon-design-system/carbon/pull/12000
2. https://github.com/carbon-design-system/carbon/pull/12222

A huge thanks goes out to @mbarrer for this work! 🙏 

Feedback on these two approaches would be appreciated and can be placed in the thread of this message and/or on the PRs themselves.


### Approach 1

This is the initial planned strategy discussed above. Definition files are moved over from DT and updated to match the current spec. The definition files live in each components' folder, are maintained separately from the component source, and are included in the package. The storybook config is updated and stories are typed as a way to validate the types.

### Approach 2

The second approach is a more straightforward refactor including a typed interface in the component source. This has some benefits over the first approach, largely that the storybook config does not change, story files need not be typed, and the types themselves live in the same file as the component source. 

### Next steps

1. Determine the best direction between these two approaches based on feedback from the community and the core team. One of the above PRs will be merged in.

2. We'll create an epic/umbrella issue to outline the work needed to provide types for each component. 

3. Organize a community effort to chew through the list of components, providing types for each one. Anyone that is interested can contribute a PR to accelerate the process of getting the types implemented for `@carbon/react`. 

As always, we'll provide updates in this discussion along the way.

**Reply by seanlaff** [reply]
With an organization as large as IBM, using typescript in the core code as in Approach 2 sounds like a great long term investment for robustness/reliability. It will make it easier for new contributors to make changes with high confidence. 👍 

Approach 1 feels like only a modest improvement over DefinitelyTyped (its in carbon instead of a third party's repo). At a glance it seems at risk for the same staleness/incorrectness issues DT suffers since the types don't live _with_ the code.

**Comment by tay1orjones** [comment]
## Status update: call for contributions! ✨ 

Hi everyone! I just want to post an update on the overall status of the effort towards TypeScript support in `@carbon/react`. @mbarrer's final piece of work to lay the foundations for TypeScript support was recently merged in 🎉 

This means **we can now begin the work to add component types to the repository**. This will be a community effort. We've organized all the work to be done within an [umbrella issue](https://github.com/carbon-design-system/carbon/issues/12513) with children issues for each component.

For anyone interested in helping out, here's a high level of how to get started:

1. [Read the docs](https://github.com/carbon-design-system/carbon/blob/main/docs/guides/adding-component-types.md) to understand the intent and scope of the effort
2. Make sure to familiarize yourself with the [steps to follow for each component](https://github.com/carbon-design-system/carbon/blob/main/docs/guides/adding-component-types.md#steps-to-provide-baseline-type-definitions-for-components)
3. [Pick a component to work on](https://github.com/carbon-design-system/carbon/issues/12513)
4. Dive in, and feel free to post any questions or suggestions to the discussion, on [discord](https://discord.com/channels/689212587170201628/1011662356977287219), or for IBMers there is an internal slack channel, [`#carbon-wg-typescript`](https://ibm-studios.slack.com/archives/C03C8VASVED)


With all this in place, we're so excited for this community effort to really push forward in 2023! Happy typing, and thanks to everyone who has pitched in so far to get this to where it’s at today! 🙏

**Comment by mthaak** [comment]
What is the status of this?

**Reply by tay1orjones** [reply]
Current status can be found in the issue:

* https://github.com/carbon-design-system/carbon/issues/12513

There's also a [TypeScript Adoption project](https://github.com/orgs/carbon-design-system/projects/53) with the same info. It contains an [insights chart](https://github.com/orgs/carbon-design-system/projects/53/insights/2) that shows progress over time.

**Comment by kurtwaldowski-b** [comment]
I know that most of the efforts here have been spent on adding types to components thus far; however, I noticed in v11 there are now javascript exports for the "elements" packages, such as color, layout, etc. 

Can we expect those to be typed at some point? Thanks!

**Reply by tay1orjones** [reply]
Not opposed to the idea but right now there are no plans to type those packages

**Comment by tay1orjones** [comment]
Hey there! We've been iterating on providing types directly through `@carbon/react` for quite some time now. We're excited to share that `v1.75.0` is the first release that contains types for every component! 🎉  With this milestone met we'll be closing this discussion.

Thank you to the many many contributors from the community who helped with this effort! 🙏 We're eager to continue working with you to improve the fidelity of types within the system and provide an excellent developer experience when working with Carbon. 

For type-related questions, concerns, or ideas moving forward, please [open an issue](https://github.com/carbon-design-system/carbon/issues/new/choose).
